numpy/core/src/multiarray/datetime.c

```diff
@@ -478,7 +491,12 @@ convert_datetime_to_datetimestruct(PyArray_DatetimeMetaData *meta,
                 dt  = dt % perday;
             }
             else {
-                set_datetimestruct_days((dt - (perday-1)) / perday, out);
+                /*
+                 * Careful with extra arithmetic on time delta subtraction
+                 * from the epoch--if dt is near the negative time span
+                 * limit it will be susceptible to overflow
+                 */
+                set_datetimestruct_days(dt / perday - (1 - 1/perday), out);
```

Comment:
The key is to avoid factoring out the common division so that `dt` is at a lower risk of overflow. If `NPY_MIN_INT64 == dt` the subtraction would overflow immediately.

Reply:
`(1 - 1/perday)` is just a weird way to spell `1`.

This gives the wrong result when `dt % perday == 0`

You might want to look at the `INT_divide` loops for how we handle this elsewhere:
```
        if (in2 == 0 || (in1 == NPY_MIN_@TYPE@ && in2 == -1)) {
            npy_set_floatstatus_divbyzero();
            *((@type@ *)op1) = 0;
        }
        else if (((in1 > 0) != (in2 > 0)) && (in1 % in2 != 0)) {
            *((@type@ *)op1) = in1/in2 - 1;
        }
        else {
            *((@type@ *)op1) = in1/in2;
        }
```

Reply:
I'd recommend introducing a simple:
```C
npy_int64 extract_unit(npy_datetime *d, npy_datetime positive_unit) {
   if (d >= 0) {
       npy_int64 div = *d / positive_unit;
       *d = *d % positive_unit;
       return div;
   }
   else {
       npy_int64 div = *d / positive_unit;
       *d = *d % positive_unit;
       if (*d != 0) {
           *d += positive_unit;
           div -= 1;
       }
       return div;
   }
}
```

Used as:
```
set_datetimestruct_days(extract_unit(&dt, perday), out);
```

Reply:
I have a few thoughts here:

1. The previous code was prone to overflow and the new version is less so, and I've added a unit test for that.
2. Are we sure there are no cases where the old code wouldn't do the wrong thing, even ignoring the overflow matter?
3. If I really am breaking something serious with this change, why is it not showing up in the unit tests? In my view, if it wasn't tested before it was already effectively broken. Can you provide a unit test to demonstrate the issue to begin with? If there's a unit test provided that probes for the matter across all possible units then I can try to fix this if there's a real problem.

Reply:
I'm quite sure this is breaking something, but I don't know enough about what hits these codepaths to construct a test.

The change you've made here assumes that `(a + b) / c == a/c + b/c`, which is not true for integral arithmetic.

I will try to come up with a test when I find the cycles. Setting dt to something like `-24H` or -1440m` should be the way to make it happen.

Agreed that if no existing tests catch this we need to add another.

Reply:
Here's your test, on your branch as of this comment:
```
In [5]: np.datetime64(-24, 'h')
Out[5]: numpy.datetime64('1969-12-30T00','h')  # incorrect

In [6]: np.datetime64(-23, 'h')
Out[6]: numpy.datetime64('1969-12-31T01','h')  # correct
```
These two ought to be an hour apart, but they're 25 hours apart

Reply:
I'll make a separate PR to apply the above cleanup, without doing any attempts at preventing overflow

Reply:
#13188